bootstrap/kubeadm/internal/webhooks/kubeadmconfig.go

```diff
@@ -0,0 +1,152 @@
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+
+	"sigs.k8s.io/cluster-api/util"
+)
+
+func (webhook *KubeadmConfig) SetupWebhookWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewWebhookManagedBy(mgr).
+		For(&bootstrapv1.KubeadmConfig{}).
+		WithDefaulter(webhook).
+		WithValidator(webhook).
+		Complete()
+}
+
+// +kubebuilder:webhook:verbs=create;update,path=/mutate-bootstrap-cluster-x-k8s-io-v1beta1-kubeadmconfig,mutating=true,failurePolicy=fail,groups=bootstrap.cluster.x-k8s.io,resources=kubeadmconfigs,versions=v1beta1,name=default.kubeadmconfig.bootstrap.cluster.x-k8s.io,sideEffects=None,admissionReviewVersions=v1;v1beta1
+// +kubebuilder:webhook:verbs=create;update,path=/validate-bootstrap-cluster-x-k8s-io-v1beta1-kubeadmconfig,mutating=false,failurePolicy=fail,matchPolicy=Equivalent,groups=bootstrap.cluster.x-k8s.io,resources=kubeadmconfigs,versions=v1beta1,name=validation.kubeadmconfig.bootstrap.cluster.x-k8s.io,sideEffects=None,admissionReviewVersions=v1;v1beta1
+
+// KubeadmConfig implements a validation and defaulting webhook for KubeadmConfig.
+type KubeadmConfig struct{}
+
+var _ webhook.CustomValidator = &KubeadmConfig{}
+var _ webhook.CustomDefaulter = &KubeadmConfig{}
+
+// Default implements webhook.Defaulter so a webhook will be registered for the type.
+func (webhook *KubeadmConfig) Default(_ context.Context, obj runtime.Object) error {
+	c, ok := obj.(*bootstrapv1.KubeadmConfig)
+	if !ok {
+		return apierrors.NewBadRequest(fmt.Sprintf("expected a KubeadmConfig but got a %T", obj))
+	}
+
+	util.DefaultKubeadmConfigSpec(&c.Spec)
+
+	return nil
+}
+
+// ValidateCreate implements webhook.Validator so a webhook will be registered for the type.
+func (webhook *KubeadmConfig) ValidateCreate(_ context.Context, obj runtime.Object) (admission.Warnings, error) {
+	c, ok := obj.(*bootstrapv1.KubeadmConfig)
+	if !ok {
+		return nil, apierrors.NewBadRequest(fmt.Sprintf("expected a KubeadmConfig but got a %T", obj))
+	}
+
+	return nil, webhook.validate(c.Spec, c.Name)
+}
+
+// ValidateUpdate implements webhook.Validator so a webhook will be registered for the type.
+func (webhook *KubeadmConfig) ValidateUpdate(_ context.Context, _, newObj runtime.Object) (admission.Warnings, error) {
+	newC, ok := newObj.(*bootstrapv1.KubeadmConfig)
+	if !ok {
+		return nil, apierrors.NewBadRequest(fmt.Sprintf("expected a KubeadmConfig but got a %T", newObj))
+	}
+
+	return nil, webhook.validate(newC.Spec, newC.Name)
+}
+
+// ValidateDelete implements webhook.Validator so a webhook will be registered for the type.
+func (webhook *KubeadmConfig) ValidateDelete(_ context.Context, _ runtime.Object) (admission.Warnings, error) {
+	return nil, nil
+}
+
+func (webhook *KubeadmConfig) validate(c bootstrapv1.KubeadmConfigSpec, name string) error {
+	allErrs := c.Validate(field.NewPath("spec"))
```

Comment:
Any reason not to have the validation in this field rather than in the API? The issue with having it in the API is it then adds code to the API and that will cause dependencies. I think the validation should all be moved over here

Reply:
Let's finish the discussion above first. But I think we can move the ValidateCreate/Update/Delete funcs without touching this code

Reply:
I think in any case it makes sense to do one PR first where we only move the funcs of the webhook interfaces. That removes the dependencies on controller-runtime. We can then discuss if we want to move more